persistence/sql/persister_nonce.go

```diff
@@ -0,0 +1,67 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+//go:generate buf generate .
+
+package sql
+
+import (
+	"context"
+	"crypto/sha256"
+	"crypto/subtle"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/ory/fosite"
+	"github.com/ory/x/errorsx"
+	"github.com/ory/x/otelx"
+)
+
+func accessTokenHash(accessToken string) []byte {
+	return sha256.New().Sum([]byte(accessToken))
+}
+
+// Set the purpose to something unique to avoid ciphertext confusion with other usages of the AEAD cipher.
+var purpose = []byte("nonce")
+
+func (p *Persister) NewNonce(ctx context.Context, accessToken string, expiresIn time.Time) (res string, err error) {
```

Comment:
I feel that generating this nonce (while a great idea in principle) could be made simpler and potentially more secure by not hashing and comparing the access token. 

Here's my take:

```go
// Copyright © 2023 Ory Corp
// SPDX-License-Identifier: Apache-2.0

package sql

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/ory/fosite"
	"github.com/ory/x/errorsx"
	"github.com/ory/x/otelx"
)

func (p *Persister) NewNonce(ctx context.Context, accessToken string, expiresIn time.Time) (res string, err error) {
	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.NewNonce")
	defer otelx.End(span, &err)

	plaintext := fmt.Sprintf("%d", expiresIn.Unix())
	aad := []byte("vc-nonce-at:" + accessToken)

	return p.r.FlowCipher().Encrypt(ctx, []byte(plaintext), aad)
}

func (p *Persister) IsNonceValid(ctx context.Context, accessToken, nonce string) (err error) {
	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.IsNonceValid")
	defer otelx.End(span, &err)

	aad := []byte("vc-nonce-at:" + accessToken)
	plaintext, err := p.r.FlowCipher().Decrypt(ctx, nonce, aad)
	if err != nil {
		return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("Invalid nonce."))
	}

	exp, err := strconv.ParseInt(string(plaintext), 10, 64)
	if err != nil {
		return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("Invalid nonce.")) // should never happen
	}

	if exp < time.Now().Unix() {
		return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("The nonce has expired."))
	}

	return nil
}
```

The protobuf solution is admittedly more easily extendable in the future.

Reply:
My reason on why to encrypt the access token was that it is unexpected that the nonce contains the access token in plain text. Clients might pass the nonce around or store them less securely than they would the access token.

The reason for protobuf over a hand-rolled format was so that we can extend what goes into the nonce without too much headache.

Reply:
The nonce we return does not contain the AAD, though. Only the base64-encoded ciphertext is returned. The AAD is only constructed on-the-fly when needed, and never exposed.

Reply:
OK got it, and I agree that this is nicer, because then the nonce is a lot smaller as well.